src/raftstore/store/worker/read.rs

```diff
@@ -289,27 +207,33 @@ impl<C: ProposalRouter> LocalReader<C> {
         cmd.callback.invoke_read(read_resp);
     }
 
-    fn pre_propose_raft_command<'a>(
-        &'a self,
-        req: &RaftCmdRequest,
-    ) -> Result<Option<&'a ReadDelegate>> {
+    fn pre_propose_raft_command(&self, req: &RaftCmdRequest) -> Result<Option<ReadDelegate>> {
         // Check store id.
-        if let Err(e) = util::check_store_id(req, self.store_id) {
+        if self.store_id.get().is_none() {
+            let store_id = self.store_meta.lock().unwrap().store_id;
+            self.store_id.set(store_id);
+        }
+        let store_id = self.store_id.get().unwrap();
+
+        if let Err(e) = util::check_store_id(req, store_id) {
             self.metrics.borrow_mut().rejected_by_store_id_mismatch += 1;
             debug!("rejected by store id not match"; "err" => %e);
             return Err(e);
         }
 
         // Check region id.
         let region_id = req.get_header().get_region_id();
-        let delegate = match self.delegates.get(&region_id) {
+        let delegate = match self.delegates.borrow_mut().get_mut(&region_id) {
             Some(delegate) => {
                 fail_point!("localreader_on_find_delegate");
-                delegate
+                match delegate.take() {
```

Comment:
Why take away? How about following read request for this region?

Reply:
It's because of lifetime. After handle a request, it will be put into the hashmap back.

Reply:
This may be a performance issue, please pay attention to it.

Reply:
Ok.